### PR TITLE
49 create design note for sequencing pv events from otel unique graphs

### PR DESCRIPTION
**Pull request by FreddieMatherSmartDCSIT**
Design note for sequencing OTel events into PV events from unique graphs calculated from OTel data.

### CODE REVIEW

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+) -> list[list[OTelEvent]]:
+    """Order the groups of otel events by their start time
+
+    :param groups: the groups of otel events to order
+    :return: the groups of otel events ordered by their start time
+    """
+
+def sequence_events_by_async_event_types(
+    events: list[OTelEvent],
+    async_event_types: list[list[str]],
```
Comment (AdrianMontaguSmartDCSIT) [inline docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md:43]: In the algo overview it's async_event_types is specified as list of sets

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    :param events: the list of otel events to sequence in order of their start time
+    :return: a list of lists of otel events with those in the same list parallel
+    """
+
+def sequence_otel_event_and_ancestors(
+    event: OTelEvent,
+    event_id_to_event_map: dict[str, OTelEvent],
+    previous_event_ids: list[str] | None = None,
+    async_flag: bool = False,
+    event_type_to_async_events_map: dict[str, list[list[str]]] | None = None,
```
Comment (AdrianMontaguSmartDCSIT) [inline docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md:66]: Algo overview specifies list of sets for event_type_to_async_events_map

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    :param event_id_to_event_map: a map of event ids to otel events
+    :param previous_event_ids: a list of previous event ids to sequence before the event
+    :param async_flag: a flag to indicate if the sequencing should be done asynchronously
+    :param event_type_to_async_events_map: a map of event types to lists of lists of event types that should be sequenced in parallel
+    """
+
+def sequence_otel_event_job(
+    job: dict[str, OTelEvent],
+    async_flag: bool = False,
+    event_type_to_async_events_map: dict[str, list[list[str]]] | None = None,
```
Comment (AdrianMontaguSmartDCSIT) [inline docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md:80]: Same as above

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    """Sequence the pv events from the otel unique graph
+
+    :param job: the otel unique graph as a dictionary of event ids to otel events
+    :param async_flag: a flag to indicate if the sequencing should be done asynchronously
+    :param event_type_to_async_events_map: a map of event types to lists of lists of event types that should be sequenced in parallel
+    :return: an Iterable of pv events sequenced from the otel unique graph
+    """
+
+
+def sequence_otel_jobs(jobs: Iterable[dict[str, OTelEvent]]) -> Iterable[Iterable[PVEvent]]:
```
Comment (AdrianMontaguSmartDCSIT) [inline docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md:91]: Algo overview has includes an async_flag

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md (hunk 1 above)
Comment (FreddieMatherSmartDCSIT) [reply docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md:43]: changed in algorithm overview

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md (hunk 4 above)
Comment (FreddieMatherSmartDCSIT) [reply docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/DN_4_Sequencing_pv_events_from_otel_unique_graphs.md:91]: updated

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/Algorithm_Overview.puml [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+            :Call "sequence_otel_event_and_ancestors";
+            repeat
+                :get OTelEvent from map of event id to OTelEvent;
+                :create PVEvent from OTelEvent using previous event ids;
+                :yield PVEvent;
+            repeat while (for each event id and previous event ids\nin map of event id to previous event ids)
+        end group
+        end
+        start
+        :Input ->\nOTelEvent,map of event id to OTelEvent,\nprevious event ids list,\nasync flag,\nmap of event type to list of lists of async following event types;
```
Comment (AdrianMontaguSmartDCSIT) [inline docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/Algorithm_Overview.puml:27]: Should this input be the same as above?
```suggestion
        group "sequence_otel_event_job"
            :find root OTelEvent;
            :Input ->\nroot OTelEvent,\nmap of event_id to OTelEvent,\nasync flag,\nmap of event type to list of lists of async following event types;
```

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/Algorithm_Overview.puml [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+        group "sequence_otel_event_and_ancestors"
+            :create empty map of event id to previous event ids;
+            :get list of child OTelEvents from event_id to OTelEvent map;
+            if (OTelEvent event type is in map of event type\nto list of sets of async following event types\nand async flag set) then (true)
```
Comment (AdrianMontaguSmartDCSIT) [inline docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/Algorithm_Overview.puml:31]:
```suggestion
            if (OTelEvent event type is in map of event type\nto list of lists of async following event types\nand async flag set) then (true)
```

File: docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/Algorithm_Overview.puml (hunk 5 above)
Comment (FreddieMatherSmartDCSIT) [reply docs/development/design/4-DN-Sequencing_pv_events_from_otel_unique_graphs/Algorithm_Overview.puml:27]: No that previous event ids list is required

**Review by AdrianMontaguSmartDCSIT — APPROVED** [review]
Looks good!

**Pull request merged**